rep-0149.rst

```diff
@@ -0,0 +1,973 @@
+  in <font color="red">red</font>.
+  Trivial changes appear in <font color="green">green</font>.
+
+
+Abstract
+========
+
+.. raw:: html
+
+  <font color="green">
+
+This REP specifies the third ``package.xml`` format, replacing the ones
+specified in REP-0127 [1]_ and REP-0140 [2]_.
+It is relevant for packages using the catkin as well as the ament build system.
+
+.. raw:: html
+
+  </font>
+
+
+Motivation
+==========
+
+.. raw:: html
+
+  <font color="red">
+
+This REP covers three separate topics which are described in the following
+subsections.
+
+Dependency groups
+-----------------
+
+Several ROS packages aggregate dependencies of a specific type, e.g.
+`message_generation` in ROS 1 or `rmw_implementation` in ROS 2.
+They explicitly list the dependencies which are being used during the release
+process.
+But when building from source they need to be updated to add custom message
+generators or rmw implementations.
+
+The goal is to allow specifying this kind of group dependency in a way that a
+from source build can process the custom packages accordingly to make them
+available to the package even though the package doesn't list an explicit
+dependency on the custom package.
+
+ABI version on the package level
+--------------------------------
+
+When building binary packages on the ROS buildfarm there is no assumption
+about ABI compatibility.
+When a new Debian package has been built all downstream packages are being
+rebuild too in order to ensure they still work together.
+This makes e.g. releasing a core package like `catkin` or `roscpp`
+particularly costly.
+Due to that overhead changes which should be released rather sooner than
+later, e.g. a documentation fix, simple bugfix in a `.cpp` file, are being
+hold back to group them with other changes to make the rebuild effort
+worthwhile.
+
+In order to allow package releases without enforcing a rebuild of all
+downstream packages a package should be able to declare if such a rebuild
+is necessary or not.
+
+Dependencies across ROS 1 and ROS 2
+-----------------------------------
+
+In order for a ROS package to work with ROS 1 and ROS 2 from a single source
+the dependencies listed in the package manifest are a problem.
+E.g. in ROS 1 a package needs to depend on `roscpp` where as in ROS 2 it needs
+to depend on `rclcpp`.
+Currently the manifest doesn't support this use case.
+
+.. raw:: html
+
+  </font>
+
+
+Rationale
+=========
+
+.. raw:: html
+
+  <font color="green">
+
+REP-0127 [1]_ and REP-0140 [1]_ provide the ``package.xml`` design rationale
+for format one and two, which is not repeated here.
+
+.. raw:: html
+
+  </font>
+
+.. raw:: html
+
+  <font color="red">
+
+Group dependencies
+------------------
+
+While "normal" dependencies are specified in the package which expresses a
+dependency on another package this is not adequate for the targeted goal.
+Instead packages need to be able to declare that they are a dependency of
+another package.
```

Comment:
This paragraph is hard to read, I'm not sure how to change it but, unless I misinterpret its meaning, I think this is clearer:

> Packages can already express direct dependencies on other individual packages, but this is not sufficient for cases where a package needs to be built after all instances of a particular kind of package.
> Instead packages need to be able to declare a dependency on packages without using their name directly.

Also, it seems to still be worded as if we're doing reverse dependency declarations, which group dependencies are sort of like, but it's a strange way to describe them.

Reply:
It seems like you're describing reverse dependencies as a discarded alternative, but I think it's confusing as presented.